### PR TITLE
Watch bookmarks may contain version of objects of other types

### PR DESCRIPTION
**Pull request by wojtek-t**
Follow up from https://github.com/kubernetes/kubernetes/pull/94364

Ref https://github.com/kubernetes/enhancements/issues/1904

/kind feature

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 1]
```diff
@@ -388,20 +388,22 @@ func (w *watchCache) UpdateResourceVersion(resourceVersion string) {
...
-	//
-	// TODO: If at some point we decide the performance and scalability
-	// footprint is acceptable, this is the place to hook them in.
-	// However, we then need to check if this was called as a result
-	// of a bookmark event or regular Add/Update/Delete operation by
-	// checking if resourceVersion here has changed.
+	go func() {
+		w.Lock()
+		defer w.Unlock()
+		w.resourceVersion = rv
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:394]: How does this prevent out of order execution? The prior code arguably might have been serialized but this code could end up with e.g. two copies of this goroutine both waiting on the lock, then the "wrong" one gets it first.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 2]
```diff
@@ -388,20 +388,22 @@ func (w *watchCache) UpdateResourceVersion(resourceVersion string) {
...
+	}()
+
+	// Avoid calling event handler under lock.
+	// This is safe as long as there is at most one call to processEvent in flight
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:398]: which processEvent? The one in this file? The relation between that function and this one isn't obvious, so it's really hard to be sure that that condition holds.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:394]: This is relying on the reflector implementation that all Add/Update/Delete and UpdateResourceVersion are happening synchronously:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L493

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:398]: Copy-paste issue - will update.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:395]: Updated the comment to reflect that.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:395]: I can believe it for the below section, but not for this bit running in a goroutine.

1. Imagine 3 successive calls A, B, C resulting in a run of this goroutine.
2. Call A happens first, takes the lock, and then the CPU context-switches away.
3. Then call B starts and blocks waiting for the lock.
4. Then call C starts and blocks waiting for the lock.
5. Then call A gets another timeslice, completes, and releases the lock.
6. Then call **C** acquires the lock and sets RV
7. Then call **B** acquires the lock and sets RV, making it go back in time.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:395]: OOps - thanks for catching this bug.
It wasn't supposed to be goroutine - it was supposed to be just func to lock only that instruction.

Fixed now - PTAL

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:395]: Ah that makes more sense :)

**Pull request merged**